File: haohaninfo/即時範例/71.py
# -*- coding: UTF-8 -*-

#取得報價資訊，詳情請查看技巧51
execfile('function.py')

#定義判斷時間
trendTime0=datetime.datetime.strptime('08:45:00.00',"%H:%M:%S.%f")
trendTime1=datetime.datetime.strptime('08:50:00.00',"%H:%M:%S.%f")
trendTime2=datetime.datetime.strptime('08:55:00.00',"%H:%M:%S.%f")
trendTime3=datetime.datetime.strptime('09:00:00.00',"%H:%M:%S.%f")
trendNum=0
trend=0

#定義指標變數
lastBAmount=0
lastSAmount=0

#取得委託資訊
for i in getOrder():		
 OrderInfo=i.split(',')
 OrderTime=datetime.datetime.strptime(OrderInfo[0],"%H:%M:%S.%f")
 OrderBAmount=int(OrderInfo[2])
 OrderSAmount=int(OrderInfo[4])

 if OrderTime>=trendTime0 and lastBAmount==0 and lastSAmount==0:
  lastBAmount=OrderBAmount
  lastSAmount=OrderSAmount

 #趨勢判斷
 if OrderTime>=trendTime1 and trendNum==0:
  diffBAmount=OrderBAmount-lastBAmount
  diffSAmount=OrderSAmount-lastSAmount
  if diffBAmount > diffSAmount:
   trend+=1
  elif diffBAmount < diffSAmount:  	
   trend-=1
  trendNum+=1
  lastBAmount=OrderBAmount
  lastSAmount=OrderSAmount
  print OrderInfo[0],"B",diffBAmount,"S",diffSAmount

 #趨勢判斷
 if OrderTime>=trendTime2 and trendNum==1:
  diffBAmount=OrderBAmount-lastBAmount
  diffSAmount=OrderSAmount-lastSAmount
  if diffBAmount > diffSAmount:
   trend+=1
  elif diffBAmount < diffSAmount:   
   trend-=1
  trendNum+=1
  lastBAmount=OrderBAmount
  lastSAmount=OrderSAmount
  print OrderInfo[0],"B",diffBAmount,"S",diffSAmount

 #趨勢判斷
 if OrderTime>=trendTime2 and trendNum==2:
  diffBAmount=OrderBAmount-lastBAmount
  diffSAmount=OrderSAmount-lastSAmount
  if diffBAmount > diffSAmount:
   trend+=1
  elif diffBAmount < diffSAmount:   
   trend-=1
  print OrderInfo[0],"B",diffBAmount,"S",diffSAmount
  break 

print "Trend",trend

File: haohaninfo/交易實務案例/71.R
#取得即時報價，詳細在技巧51
source("function.R")

#三個判斷區段，設定四個判斷時間點
trendTime0 <- strptime('08:45:00.00','%H:%M:%OS')
trendTime1 <- strptime('08:50:00.00','%H:%M:%OS')
trendTime2 <- strptime('08:55:00.00','%H:%M:%OS')
trendTime3 <- strptime('09:00:00.00','%H:%M:%OS')
#紀錄每上個時間點的總量
lastBuyAmount <- 0
lastSellAmount <- 0
#用變數作為趨勢初始值
trend <- 0
#用另一個變數來記錄目前的趨勢判斷進度，避免重複判斷
trendnum <- 0

while(TRUE){

 Odata<-GetOrderData(DataPath,Date)
 #因為要進行比較，所以透過R內建函數strptime轉為R語言的時間格式
 #最小單位可至秒數以下兩位
 OrderTime <- strptime(Odata[[1]][1],'%H:%M:%OS')
 OrderBuyAmount <- as.numeric(Odata[[1]][3])
 OrderSellAmount <- as.numeric(Odata[[1]][5])

 #進行第一次的總量紀錄
 if ( OrderTime > trendTime0 & lastBuyAmount == 0){
  lastBuyAmount <- OrderBuyAmount
  lastSellAmount <- OrderSellAmount
 }

 #第一次判斷委託變動量
 if( OrderTime > trendTime1 & trendnum == 0 ){
  #計算8:45至8:50委託總量差異
  diffBuyAmount <- OrderBuyAmount - lastBuyAmount
  diffSellAmount <- OrderSellAmount - lastSellAmount

  if( diffBuyAmount > diffSellAmount ){
   trend <- trend + 1
  }else if( diffBuyAmount < diffSellAmount ){
   trend <- trend - 1
  }
  trendnum <- trendnum +1
  lastBuyAmount <- OrderBuyAmount
  lastSellAmount <- OrderSellAmount
 }

 #第二次判斷委託變動量
 if( OrderTime > trendTime2 & trendnum == 1 ){
  #計算8:50至8:55委託總量差異 
  diffBuyAmount <- OrderBuyAmount - lastBuyAmount
  diffSellAmount <- OrderSellAmount - lastSellAmount

  if( diffBuyAmount > diffSellAmount ){
   trend <- trend + 1
  }else if( diffBuyAmount < diffSellAmount ){
   trend <- trend - 1
  }
  trendnum <- trendnum +1
  lastBuyAmount <- OrderBuyAmount
  lastSellAmount <- OrderSellAmount
 }

 #第三次判斷委託變動量
 if( OrderTime > trendTime3 & trendnum == 2 ){
  #計算8:55至9:00委託總量差異
  diffBuyAmount <- OrderBuyAmount - lastBuyAmount
  diffSellAmount <- OrderSellAmount - lastSellAmount

  if( diffBuyAmount > diffSellAmount ){
   trend <- trend + 1
  }else if( diffBuyAmount < diffSellAmount ){
   trend <- trend - 1
  }
  trendnum <- trendnum +1
  lastBuyAmount <- OrderBuyAmount
  lastSellAmount <- OrderSellAmount
 }


 #判斷完成後，顯示預測趨勢
 #若trend大於0，則看多；小於0，則看空
 if( trendnum == 3 ){
  print(trend)
 }

}
